### ISSUE TITLE
Valid "unreachable" code is being aggressively removed as a result of syntax errors when allowUnreachableCode: false

### ISSUE THREAD

**Issue by bbugh** [microsoft/vscode#109530]
- VSCode Version: 1.50.1
- OS Version: 10.14.6
- TypeScript: 4.0.3

Steps to Reproduce:

1. Set `allowUnreachableCode: false` in tsconfig
1. In a TS file, have a syntax error that TypeScript thinks results in unreachable code (see screenshot)
1. Save a TS file
1. Watch as valid code is removed, causing a lot of confusion about why everything breaks suddenly

Does this issue occur when all extensions are disabled?: Yes

This has been happening a lot lately but as far as I can remember it's never happened before. We had recently upgraded TypeScript from 3.7.2 to 4.0.3, but as far as I can tell with investigation nothing else about our app configuration or VSCode common settings have changed.

I disabled all extensions, ran all of the linters separately, etc, and my code keeps being removed when I accidentally save with syntax errors. 

The TypeScript compiler reports the syntax issues but does not report a `7027` error, the unreachable code error only appears in VS Code.

This does not happen when `allowUnreachableCode` is `true`, but we don't want to allow unreachable code.

Example: 

![Pasted_Image_10_27_20__2_31_PM](https://user-images.githubusercontent.com/438465/97352891-3e3cd300-1861-11eb-90cc-44223cf8653f.png)

If I save this file, the `underscore` code will be removed.

<img width="515" alt="image" src="https://user-images.githubusercontent.com/438465/97352742-0c2b7100-1861-11eb-970b-d5b7b08e2fb2.png">

**Comment by mjbvz** [comment]
Please always share the text of the code along with screenshots

**Comment by bbugh** [comment]
Sure. I didn't type it out because that code doesn't matter, it's any unreachable code. Here's a minimally reproducible file. Copy/paste, then save, and the `console.log` line will disappear.

```typescript
// some_file.ts
return

console.log("hello")
```

Type it out, Cmd+S, voila!

![Screen-Recording-2020-10-27-at-4](https://user-images.githubusercontent.com/438465/97363992-5fa5bb00-1871-11eb-8774-cd11a49a94b0.gif)

**Comment by mjbvz** [comment]
This is not the default behavior. It looks like you've configured `code actions on save`, which means this is the expected behavior

We generally can't know that the code is unreachable because of an expected syntax error vs being truly unreachable (and we also can't assume that code will always be syntax error free and just disable unreachable code checking)

**Comment by bbugh** [comment]
I am not concerned about the setting value, I am reporting that the behavior of the "all inclusive" settings is breaking things in a really bad way. I'm very confused how you are suggesting that this is acceptable behavior for a code editor. The demo video shows how much of a drastic breaking change it is to code.

**Comment by matt-tingen** [comment]
For what it's worth, I also found this behavior to be quite surprising and it took a few instances for me to be sure I wasn't accidentally deleting code myself.

I didn't even realize TS had fixes to run so I was using `source.fixAll` on save thinking eslint was the only provider. Changing that to `source.fixAll.eslint` works for me, but I figured I should share my experience.

As a point of comparison, it's worth mentioning that eslint's [`no-unreachable`](https://eslint.org/docs/rules/no-unreachable) does not have an auto-fix and that's on top of eslint bailing out when there's a syntax error.
